File: day-2-pm/sandbox/tests/evaluator_test.cpp
#include "evaluator.h"

#include <gtest/gtest.h>

#include "parser.h"

namespace {

std::int64_t eval_str(const char* src) {
    auto ast = calc::parse(src);
    return calc::evaluate(*ast);
}

}  // namespace

TEST(EvaluatorTest, Addition) { EXPECT_EQ(eval_str("1+2"), 3); }

TEST(EvaluatorTest, Multiplication) { EXPECT_EQ(eval_str("3*4"), 12); }

TEST(EvaluatorTest, IntegerDivisionTruncates) {
    EXPECT_EQ(eval_str("10/2"), 5);
    EXPECT_EQ(eval_str("7/2"), 3);
}

TEST(EvaluatorTest, Subtraction) { EXPECT_EQ(eval_str("5-3"), 2); }

TEST(EvaluatorTest, MixedPrecedence) {
    EXPECT_EQ(eval_str("1+2*3"), 7);
    EXPECT_EQ(eval_str("2*3+4"), 10);
    EXPECT_EQ(eval_str("20-6/2"), 17);
}

TEST(EvaluatorTest, DivisionByZeroThrows) {
    EXPECT_THROW(eval_str("1/0"), std::runtime_error);
}

File: day-2-pm/sandbox-java/src/test/java/calc/EvaluatorTest.java
package calc;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;

import org.junit.jupiter.api.Test;

class EvaluatorTest {

    private static long evalStr(String src) {
        Node ast = Parser.parse(src);
        return Evaluator.evaluate(ast);
    }

    @Test
    void Addition() {
        assertEquals(3, evalStr("1+2"));
    }

    @Test
    void Multiplication() {
        assertEquals(12, evalStr("3*4"));
    }

    @Test
    void IntegerDivisionTruncates() {
        assertEquals(5, evalStr("10/2"));
        assertEquals(3, evalStr("7/2"));
    }

    @Test
    void Subtraction() {
        assertEquals(2, evalStr("5-3"));
    }

    @Test
    void MixedPrecedence() {
        assertEquals(7, evalStr("1+2*3"));
        assertEquals(10, evalStr("2*3+4"));
        assertEquals(17, evalStr("20-6/2"));
    }

    @Test
    void DivisionByZeroThrows() {
        assertThrows(CalcException.class, () -> evalStr("1/0"));
    }
}
